### ISSUE TITLE
test.cpp file execution: all tests fails

### ISSUE THREAD

**Issue by raphael10-collab** [guitargeek/XGBoost-FastForest#24]
(.fastforest) (base) raphy@raohy:~/FastForestTest$ nano CMakeLists.txt

                                                                                
    cmake_minimum_required(VERSION 3.15...3.31)
    project(FastForestTest)
    
    #----------------------------------------------------------------------------------------------------------------------
    # testing framework
    #----------------------------------------------------------------------------------------------------------------------
    
    include(FetchContent)
    FetchContent_Declare(googletest URL https://github.com/google/googletest/archive/refs/tags/release-1.12.1.tar.gz)
    
    set(gtest_force_shared_crt ON CACHE BOOL "" FORCE) # do not override parent project's runtime settings on Windows
    set(INSTALL_GTEST OFF)
    
    # For simplicity, always build googletest as static library. This prevents fastforest-tests executable from
    # complaining about missing googletest DLLs on Windows.
    set(BUILD_SHARED_LIBS OFF)
    
    FetchContent_MakeAvailable(googletest)
    
    
    add_executable(${PROJECT_NAME}
        #./src/fastforestUsageExample.cpp
        ./src/test.cpp
    )
    
    target_link_directories(${PROJECT_NAME} PUBLIC
        ./src/FastForest/lib
    )
    
    target_include_directories(${PROJECT_NAME} PUBLIC
    
        ./src/FastForest/include
    
    )
    
    target_link_libraries (${PROJECT_NAME} PUBLIC
        fastforest
        gtest_main
    )
    
    include(GoogleTest)

The test file ./src/test.cpp  is taken  from: https://github.com/guitargeek/XGBoost-FastForest/blob/master/test/test.cpp

Executing:

    (.fastforest) (base) raphy@raohy:~/FastForestTest$ ./builddir/FastForestTest 
    Running main() from /home/raphy/FastForestTest/builddir/_deps/googletest-src/googletest/src/gtest_main.cc
    [==========] Running 8 tests from 1 test suite.
    [----------] Global test environment set-up.
    [----------] 8 tests from FastForest
    [ RUN      ] FastForest.Example
    unknown file: Failure
    C++ exception with description "constructing FastForest from continuous/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.Example (0 ms)
    [ RUN      ] FastForest.Basic
    unknown file: Failure
    C++ exception with description "constructing FastForest from continuous/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.Basic (0 ms)
    [ RUN      ] FastForest.Softmax
    unknown file: Failure
    C++ exception with description "constructing FastForest from softmax/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.Softmax (0 ms)
    [ RUN      ] FastForest.SoftmaxNSamples100NFeatures100
    unknown file: Failure
    C++ exception with description "constructing FastForest from softmax_n_samples_100_n_features_100/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.SoftmaxNSamples100NFeatures100 (0 ms)
    [ RUN      ] FastForest.SoftmaxArray
    unknown file: Failure
    C++ exception with description "constructing FastForest from softmax/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.SoftmaxArray (0 ms)
    [ RUN      ] FastForest.Serialization
    unknown file: Failure
    C++ exception with description "constructing FastForest from continuous/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.Serialization (0 ms)
    [ RUN      ] FastForest.Discrete
    unknown file: Failure
    C++ exception with description "constructing FastForest from discrete/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.Discrete (0 ms)
    [ RUN      ] FastForest.ManyFeatures
    unknown file: Failure
    C++ exception with description "constructing FastForest from manyfeatures/model.txt: file does not exists" thrown in the test body.
    [  FAILED  ] FastForest.ManyFeatures (0 ms)
    [----------] 8 tests from FastForest (0 ms total)
    
    [----------] Global test environment tear-down
    [==========] 8 tests from 1 test suite ran. (0 ms total)
    [  PASSED  ] 0 tests.
    [  FAILED  ] 8 tests, listed below:
    [  FAILED  ] FastForest.Example
    [  FAILED  ] FastForest.Basic
    [  FAILED  ] FastForest.Softmax
    [  FAILED  ] FastForest.SoftmaxNSamples100NFeatures100
    [  FAILED  ] FastForest.SoftmaxArray
    [  FAILED  ] FastForest.Serialization
    [  FAILED  ] FastForest.Discrete
    [  FAILED  ] FastForest.ManyFeatures

**Comment by guitargeek** [comment]
Thanks for the report! Which xgboost version are you using? It's very likely that things don't work with xgboost 3 yet. I will need to make an effort to support it

**Comment by raphael10-collab** [comment]
Hi  Jonas @guitargeek 

yes it's xgboost 3 : 

    (.fastforest) (base) raphy@raohy:~/XGBoost-FastForest$ pip show xgboost
    Name: xgboost
    Version: 3.0.4
    Summary: XGBoost Python Package
    Home-page: 
    Author: 
    Author-email: Hyunsu Cho <chohyu01@cs.washington.edu>, Jiaming Yuan <jm.yuan@outlook.com>
    License: Apache-2.0
    Location: /home/raphy/XGBoost-FastForest/.fastforest/lib/python3.12/site-packages
    Requires: numpy, nvidia-nccl-cu12, scipy
    Required-by:

**Comment by guitargeek** [comment]
Thanks! I have seen in a CI trial run that they have changed the implied constant in the output once again 😆 

  * [defaultBaseResponse](https://github.com/guitargeek/XGBoost-FastForest/pull/26/

So if you set the `defaultBaseResponse` parameter in the evaluation function to 0.5, it will work. I'm seeing what I can do to automatically detect the right base response based on the xgboost version. I'm not sure though if the version information is available in the `txt` dump of the model... So maybe I have to fall back to just explaining this caveat in the README. I'll keep you updated in this issue thread!

**Comment by guitargeek** [comment]
The tests now pass also with XGBoost 3 (see the latest CI run). The caveat is that you have to explicitly add the `base_score` to the text dump, but this is also explained in the README now:

  * https://github.com/guitargeek/XGBoost-FastForest?tab=readme-ov-file#usage-example
